Kubelet & API changes for Windows GMSA support

wk8 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This patch comprises the kubelet & API changes outlined in the Windows GMSA KEP
(https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20181221-windows-group-managed-service-accounts-for-container-identity.md)
to add GMSA support to Windows workloads.

This essentially means moving the GMSA cred spec names & contents
from annotations to dedicated API fields.
More specifically, this is affecting `PodSecurityContext` and `SecurityContext`
structs, adding a `WindowsOptions` object to both.

Updated tests.

**Special notes for your reviewer**:
Still a WIP, will also update the webhook
(https://github.com/kubernetes-sigs/windows-gmsa)
to ensure all is in order before merging.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
API changes and deprecating the use of special annotations for Windows GMSA support (version beta)
```

wk8 commented:
Putting on hold until the webhook is upated too. /hold

wk8 commented:
/hold

wk8 commented:
/sig-windows

wk8 commented:
/sig windows

ddebroy commented on pkg/features/kube_features.go:
```diff
@@ -506,7 +507,7 @@ var defaultKubernetesFeatureGates = map[utilfeature.Feature]utilfeature.FeatureS
 	TTLAfterFinished:                            {Default: false, PreRelease: utilfeature.Alpha},
 	KubeletPodResources:                         {Default: false, PreRelease: utilfeature.Alpha},
-	WindowsGMSA:                                 {Default: false, PreRelease: utilfeature.Alpha},
+	WindowsGMSA:                                 {Default: false, PreRelease: utilfeature.Beta},
```
Let's move this out of this PR for now. Once the overall code changes have been reviewed and all desired tests are passing, we can consider the feature bump a little later in v1.15.

wk8 replied:
👍

ddebroy reviewed (commented):
Overall looks good. Once comment about removing the feature flag. While the link to the KEP is useful, can you update the PR description please with the API objects changed: `PodSecurityContext` and `SecurityContext`) and the new fields introduced?

ddebroy commented:
/assign @liggitt @yujuhong

ddebroy commented:
Looks like the following will need to be run:
https://github.com/kubernetes/kubernetes/blob/master/hack/update-generated-protobuf.sh
https://github.com/kubernetes/kubernetes/blob/master/hack/update-generated-swagger-docs.sh
https://github.com/kubernetes/kubernetes/blob/master/hack/update-openapi-spec.sh
and the changes need to be part of the PR.

wk8 commented:
/label api-review

michmike commented:
/milestone v1.15
/sig windows

ddebroy commented:
/retest

wk8 commented:
/test pull-kubernetes-integration

liggitt commented:
/label api-review

wk8 commented:
/label api-review

wk8 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

ddebroy commented on pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:
```diff
@@ -104,39 +104,28 @@ func (m *kubeGenericRuntimeManager) generateWindowsContainerConfig(container *v1
-// GMSASpecContainerAnnotationKey annotation.
+// determineEffectiveSecurityContext determines the effective GMSA credential spec and, if any, copies it
+// to the config's `.Windows.SecurityContext.GmsaCredentialSpec`.
 func determineEffectiveSecurityContext(config *runtimeapi.ContainerConfig, container *v1.Container, pod *v1.Pod) {
```
Will it be possible to get rid of the `determineEffectiveSecurityContext` here and transfer the logic below to `securitycontext.DetermineEffectiveSecurityContext` above?

ddebroy commented:
Combined KEP for the API change related to GMSA and RunAsUsername: https://github.com/kubernetes/enhancements/pull/972

wk8 commented:
/hold cancel

wk8 commented:
@ddebroy @liggitt : rebased, amended, and did 3 separate commits (API changes, auto-gen, implementation) as requested. Could you please have another look? Thanks!

wk8 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

wk8 commented:
/test pull-kubernetes-e2e-gce-100-performance

ddebroy commented on pkg/apis/core/types.go:
```diff
@@ -4730,7 +4730,15 @@ type SELinuxOptions struct {
+
+	// GMSACredentialSpec is where the GMSA admission webhook
+	// (https://github.com/kubernetes-sigs/windows-gmsa) inlines the contents of the
+	// GMSA credential spec named by the `GmsaCredentialSpecName` field.
```
nit: Can you remove the single quotes around the `GmsaCredentialSpecName` in the comments?

ddebroy commented on pkg/kubelet/dockershim/docker_container_windows.go:
```diff
@@ -54,7 +53,7 @@ func (ds *dockerService) applyPlatformSpecificDockerConfig(request *runtimeapi.C
 }
 
-// applyGMSAConfig looks at the kuberuntime.GMSASpecContainerAnnotationKey container annotation; if present,
+// applyGMSAConfig looks at the container's `.Windows.SecurityContext.GmsaCredentialSpec` field; if present,
```
nit: I think the single quotes around `.Windows.SecurityContext.GmsaCredentialSpec` can be removed from the comments.

ddebroy reviewed (commented):
A couple of minor nits around single quotes in comments. I think those are unnecessary. Otherwise lgtm.

wk8 commented:
@ddebroy : thanks for the review, amended as requested.

wk8 commented:
/test pull-kubernetes-e2e-gce-100-performance

ddebroy commented on pkg/features/kube_features.go:
```diff
@@ -425,6 +425,7 @@ const (
 
 	// owner: @wk8
 	// alpha: v1.14
+	// beta: v1.15
```
This change is no longer true. This should stay as is.

wk8 replied:
👍

ddebroy reviewed (commented):
I have another nit. Also, I think (but not sure) there needs to be logic similar to https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L384 which checks for the feature gates around GMSA and sets the fields to nil as part of `DropDisabledPodFields` invoked through PrepareForCreate/PrepareForUpdate paths.

wk8 commented:
/test pull-kubernetes-e2e-gce-100-performance

wk8 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5187,6 +5190,38 @@ func ValidateSecurityContext(sc *core.SecurityContext, fldPath *field.Path) fiel
+		return allErrs
+	}
+
+	if windowsOptions.GMSACredentialSpecName != "" {
```
once these change to `*string`, ignore if nil, require to be a valid value if non-nil

wk8 replied:
👍

liggitt commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:
```diff
@@ -637,6 +637,9 @@ message WindowsContainerSecurityContext {
     string run_as_username = 1;
+
+    // The contents of the GMSA credential spec to use to run this container.
+    string gmsa_credential_spec = 2;
```
to align with the underlying runtime field this will get plumbed into, let's call this `credential_spec`

see https://github.com/opencontainers/runtime-spec/blob/5b71a03e2700cbb1091b78849b8748f5e1692650/schema/config-windows.json#L80

`pod.spec.securityContext.windowsOptions.gmsaCredentialSpec` is one possible source for this, but if we added more sources in the future, we wouldn't want to have to plumb all the sources through all the layers

wk8 replied:
👍

liggitt commented:
> Also, I think (but not sure) there needs to be logic similar to https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L384 which checks for the feature gates around GMSA and sets the fields to nil as part of DropDisabledPodFields invoked through PrepareForCreate/PrepareForUpdate paths.

that's exactly right. please add a corresponding unit test when you make that change (there are examples for all the other conditional fields in util_test.go in that package

liggitt commented on pkg/securitycontext/util.go:
```diff
@@ -63,6 +63,11 @@ func DetermineEffectiveSecurityContext(pod *v1.Pod, container *v1.Container) *v1
 
+	if containerSc.WindowsOptions != nil {
+		effectiveSc.WindowsOptions = new(v1.WindowsSecurityContextOptions)
+		*effectiveSc.WindowsOptions = *containerSc.WindowsOptions
```
do you expect the container windowsOptions to fully replace the pod-level windowsOptions, or if both are non-nil, should only the non-nil container windowsOptions fields overwrite the pod-level ones?

wk8 replied:
I think it makes more sense to fully replace the pod-level one; it makes little sense to set GMSA names and contents in 2 different places - although arguably this might change as we add more options. Besides, that's the current behaviour for eg SELinuxOptions, seems better to keep the same behaviour?

liggitt replied:
will that still be true as future fields are added to `windowsOptions` (e.g. runAsUsername)?

wk8 replied:
Fair enough, changing this.

liggitt replied:
I could see both gmsa fields getting overridden together (if either is non-nil in the container securityContext, both get used)

wk8 replied:
Done

wk8 commented on pkg/apis/core/types.go:
```diff
@@ -4730,7 +4730,15 @@ type SELinuxOptions struct {
 type WindowsSecurityContextOptions struct {
-	// intentionally left empty for now
+	// GMSACredentialSpecName is the name of the GMSA credential spec to use.
+	// +optional
```
👍

wk8 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5187,6 +5190,38 @@ func ValidateSecurityContext(sc *core.SecurityContext, fldPath *field.Path) fiel
+
+	if windowsOptions.GMSACredentialSpecName != "" {
+		// gMSACredentialSpecName must be the name of a custom resource
+		for _, msg := range customresource.NameValidationFunction(windowsOptions.GMSACredentialSpecName, false) {
```
👍

wk8 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5346,7 +5346,15 @@ type SELinuxOptions struct {
-	// intentionally left empty for now
+	// GMSACredentialSpecName is the name of the GMSA credential spec to use.
+	// +optional
+	GMSACredentialSpecName string `json:"gMSACredentialSpecName,omitempty" protobuf:"bytes,1,opt,name=gMSACredentialSpecName"`
```
👍

liggitt replied:
serialized names look good... go field names can stay `GMSA...`. sorry I wasn't clear about that

wk8 replied:
Sorry I misunderstood. Changed.

wk8 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5346,7 +5346,15 @@ type SELinuxOptions struct {
-	// intentionally left empty for now
+	// GMSACredentialSpecName is the name of the GMSA credential spec to use.
+	// +optional
+	GMSACredentialSpecName string `json:"gMSACredentialSpecName,omitempty" protobuf:"bytes,1,opt,name=gMSACredentialSpecName"`
```
👍

wk8 commented:
@liggitt : thanks for the review, amended as requested.

wk8 commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

liggitt commented on pkg/securitycontext/util.go:
```diff
@@ -66,6 +66,18 @@ func DetermineEffectiveSecurityContext(pod *v1.Pod, container *v1.Container) *v1
+		}
+		if containerSc.WindowsOptions.GmsaCredentialSpecName != nil || containerSc.WindowsOptions.GmsaCredentialSpec != nil {
+			// both GMSA fields go hand in hand
+			*effectiveSc.WindowsOptions.GmsaCredentialSpecName = *containerSc.WindowsOptions.GmsaCredentialSpecName
```
won't `*effectiveSc.WindowsOptions.GmsaCredentialSpecName` and `*effectiveSc.WindowsOptions.GmsaCredentialSpecName` NPE?

liggitt replied:
one of `*containerSc.WindowsOptions.GmsaCredentialSpecName` and `*containerSc.WindowsOptions.GmsaCredentialSpec` could be nil here and NPE

would this be better?
```go
if containerSc.WindowsOptions.GmsaCredentialSpecName != nil || containerSc.WindowsOptions.GmsaCredentialSpec != nil {
  if containerSc.WindowsOptions.GmsaCredentialSpecName == nil {
    effectiveSc.WindowsOptions.GmsaCredentialSpecName = nil
  } else {
    s := *containerSc.WindowsOptions.GmsaCredentialSpecName
    effectiveSc.WindowsOptions.GmsaCredentialSpecName = &s
  }

  if containerSc.WindowsOptions.GmsaCredentialSpec == nil {
    effectiveSc.WindowsOptions.GmsaCredentialSpec = nil
  } else {
    s := *containerSc.WindowsOptions.GmsaCredentialSpec
    effectiveSc.WindowsOptions.GmsaCredentialSpec = &s
  }
}
```

liggitt reviewed (commented):
thanks for the update. you can squash the `Review comment: renaming GMSA go variables` files into the right commits.

found a couple potential NPEs in pkg/securitycontext/util.go

the update to dropDisabledFields() (https://github.com/kubernetes/kubernetes/pull/75459#issuecomment-491432988) is still outstanding as well

looking good otherwise

wk8 committed:
API changes for Windows GMSA support

This patch comprises the API changes outlined in the Windows GMSA KEP
(https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20181221-windows-group-managed-service-accounts-for-container-identity.md)
to add GMSA support to Windows workloads.

It includes validation, as well as dropping fields if the `WindowsGMSA` feature
flag is not set, both with unit tests.

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 committed:
GRPC changes for GMSA support

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 committed:
Auto-generated changes for Windows GMSA support

This patch comprises the auto-generated changes for the API changes outlined in
the Windows GMSA KEP
(https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20181221-windows-group-managed-service-accounts-for-container-identity.md)
to add GMSA support to Windows workloads.

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 committed:
Kubelet & implementation changes for Windows GMSA support

This patch comprises the kubelet changes outlined in the Windows GMSA KEP
(https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20181221-windows-group-managed-service-accounts-for-container-identity.md)
to add GMSA support to Windows workloads.

Updated tests.

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 commented:
@liggitt @ddebroy : amended as requested, including https://github.com/kubernetes/kubernetes/pull/75459#issuecomment-491432988, could you please take another look?

liggitt commented:
/lgtm
/approve

liggitt commented:
/priority important-soon

liggitt commented:
/cc @yujuhong

for cri approval

yujuhong commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:
```diff
@@ -637,6 +637,9 @@ message WindowsContainerSecurityContext {
     string run_as_username = 1;
+
+    // The contents of the GMSA credential spec to use to run this container.
+    string credential_spec = 2;
```
The API change LGTM. We may need to change the comment to make it clear if we ever supports non-GMSA spects.

wk8 replied:
@yujuhong : sorry if I'm misunderstanding, but that field's name was requested to not include "GMSA" in https://github.com/kubernetes/kubernetes/pull/75459#discussion_r283044104, and the comment makes it clear it's for GMSA for now; so not sure what else we should do here?

liggitt replied:
I think this is fine for now. The field name matches the oci field name, and the description is reasonably consistent with the [oci doc](https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L442-L443).

PatrickLang commented:
@yujuhong - can you approve or find one from SIG-Node?

yujuhong commented:
/lgtm
/approve
